kbs/docs/initdata.md

```diff
@@ -0,0 +1,267 @@
+line breaks that do not affect the semantics but will cause the original initdata
+metadata content to change. This fact will make it difficult for us to calculate a
+determined digest from different forms of a same initdata. We need a canonicalization
+algorithm. With this algorithm, all different forms of a same initdata will be calculated
+into a same digest. The difference of forms at least include the following
+
+1. The encoding of the initdata metadata, e.g. JSON and TOML.
+2. Initdata metadata with different expressions but the same semantics, e.g. multiple line
+JSON vs one line JSON like the following
+```json
+{
+  "algorithm": "sha384",
+  "version": "0.1.0",
+  "data": {
+    "key1": "value1",
+    "key2": "value2"
+  },
+  "digest": "0d25829cc97872f0086fbb7eec3a08c0b76899986dc135e9545434ae9eba7d48196efa7323519f00384881b659c02087"
+}
+```
+and
+```json
+{"algorithm":"sha384","version":"0.1.0","data":{"key1":"value1","key2":"value2"},"digest":"0d25829cc97872f0086fbb7eec3a08c0b76899986dc135e9545434ae9eba7d48196efa7323519f00384881b659c02087"}
+```
+
+The algorithm has 2 steps as following
+
+1. Transform the original initdata metadata's `data` field into JSON.
+2. Canonicalize the JSON resulted in step 1. The canonicalization follows [RFC 8785](https://www.rfc-editor.org/rfc/rfc8785). This includes removing extra whitespace characters and ordering of object members.
+
+After performing the algorithm, the result string can be used to apply hash `algorithm` and get the `digest`.
+
+## Use cases
+
+### Confidential Containers
+
+In Confidential Containers, we need initdata to deliver the following configurations into the
+guest.
+- KBS public key and address
+- Kata agent policy
+- Configuration files for Attestation Agent, Confidential Data Hub and probably kata-agent in future.
+
+The initdata metadata will be inject into guest by kata-runtime via kata-agent's API. After receiving
+initdata metadata, kata-agent will calculate the digest it self due to the metadata, and call Attestation
+Agent's `CheckInitData()` api to check if the received initdata's integrity is aligned with the one
+inside the evidence. Next, it will use the key-value pairs recorded in the data as the file name and
+content, and place them in the `/run/confidential-containers/initdata` directory.
+
+The encoding format is TOML in order to facilitate viewing and modifying specific content.
+
+#### Fields of Data
+
+To prevent malicious host inject any files especially binaries into the TEE, only defined `key`s in
+`data` can be accepted. The defined `key`s including the following:
+- `attestation-agent.json`: The configuration file of Attestation Agent.
+- `confidential-data-hub.toml`: The configuration file of Confidential Data Hub.
+- `policy.json`: The rego policy file used by kata.
+
+any other keys inside `data` will be considered illegal and the entire initdata verification process will fail.
```

Comment:
Maybe this should also be left to the agent implementation. I'm not sure what we should do here yet, actually.

It seems a bit stricter to reject configs that don't match exactly what we expect, but that rules out the possibility of passing something (like a workload id) directly to the KBS. I'm also not sure how it would work with optional parameters. So we might just want to ignore unexpected fields. We can discuss.

Reply:
Agreed. These are better to be left to implementation PR reviews. Let me just delete this line and carry on [the thread](https://github.com/confidential-containers/trustee/pull/348#discussion_r1517078459)